# Data Science Project - Step 2: Data Loading & Cleaning

**Due: Sunday, April 19 | Points: 25**

In this step, you will:
1. Load your dataset with pandas
2. Explore and understand your data
3. Handle missing values
4. Fix data types and clean problems
5. Save a clean dataset for analysis

This builds on Step 1 — use the dataset you chose!

In [14]:
import pandas as pd
import os

# Create the folders you'll need (safe to run multiple times)
os.makedirs('data', exist_ok=True)
os.makedirs('outputs', exist_ok=True)
print("Folders ready!")
print("  Put your CSV file in the 'data/' folder, then continue.")

Folders ready!
  Put your CSV file in the 'data/' folder, then continue.


---
## Part 1: Load Your Data (5 points)
---

In [15]:
# 1a. Load your dataset
# Replace 'your_data.csv' with your actual filename
# Make sure your data file is in a 'data/' folder

df = pd.read_csv("data/mlb_stats_1901_to_2025.csv")

In [16]:
# 1b. Check basic info
print(f"Dataset shape: {df.shape[0]} rows, {df.shape[1]} columns")
print(f"\nColumn names:\n{df.columns.tolist()}")

Dataset shape: 2690 rows, 16 columns

Column names:
['team_name', 'year', 'wins', 'losses', 'winning_percentage', 'games_behind', 'wild_card_games_behind', 'record_in_the_last_10_games', 'current_streak', 'runs_scored', 'runs_allowed', 'run_differential', 'expected_win_loss_record', 'record_at_home', 'record_when_away', 'record_against_top_50_percent']


In [17]:
# 1c. View first few rows
df.head()

,team_name,year,wins,losses,winning_percentage,games_behind,wild_card_games_behind,record_in_the_last_10_games,current_streak,runs_scored,runs_allowed,run_differential,expected_win_loss_record,record_at_home,record_when_away,record_against_top_50_percent
0,Pittsburgh Pirates,1901,90,49,0.647,NaN,NaN,4-Jun,W1,776,534,242,92-47,45-24,45-25,47-32
1,Chicago White Sox,1901,83,53,0.610,5.5,NaN,5-May,W1,819,631,188,84-52,49-21,34-32,44-34
2,Philadelphia Phillies,1901,83,57,0.593,7.5,NaN,3-Jul,L1,668,543,125,83-57,46-23,37-34,38-42
3,Boston Americans,1901,79,57,0.581,9.5,NaN,3-Jul,W6,759,608,151,82-54,49-20,30-37,40-38
4,Brooklyn Superbas,1901,79,57,0.581,9.5,NaN,3-Jul,W3,744,600,144,81-55,43-25,36-32,41-38


In [18]:
# 1d. Check data types
df.dtypes

team_name                            str
year                               int64
wins                               int64
losses                             int64
winning_percentage               float64
games_behind                     float64
wild_card_games_behind           float64
record_in_the_last_10_games          str
current_streak                       str
runs_scored                        int64
runs_allowed                       int64
run_differential                   int64
expected_win_loss_record             str
record_at_home                       str
record_when_away                     str
record_against_top_50_percent        str
dtype: object

---
## Part 2: Explore Your Data (5 points)
---

Now that it's loaded, let's understand it better.

In [19]:
# 2a. Summary statistics for numeric columns
df.describe()

,year,wins,losses,winning_percentage,games_behind,wild_card_games_behind,runs_scored,runs_allowed,run_differential
count,2690.000000,2690.000000,2690.000000,2690.000000,2559.000000,662.000000,2690.000000,2690.000000,2690.000000
mean,1972.597026,77.330112,77.331599,0.499965,23.206331,15.358006,685.453903,685.453903,0.000000
std,35.614727,15.267892,15.212028,0.082441,13.110205,9.832843,122.977561,124.718269,118.025317
min,1901.000000,10.000000,17.000000,0.167000,0.500000,0.500000,190.000000,196.000000,-349.000000
25%,1944.000000,68.000000,68.000000,0.444000,13.000000,7.500000,623.000000,618.000000,-80.000000
50%,1978.000000,79.000000,78.000000,0.505500,22.000000,14.000000,693.000000,691.000000,4.000000
75%,2003.000000,88.000000,88.000000,0.559000,31.750000,22.000000,759.000000,762.000000,80.750000
max,2025.000000,116.000000,121.000000,0.763000,68.000000,52.000000,1067.000000,1199.000000,411.000000


In [20]:
# 2b. Value counts for categorical columns
# Pick 2-3 categorical columns and check their values
df['wins'].value_counts()



wins
86     85
83     82
75     78
76     76
79     74
       ..
114     1
24      1
19      1
18      1
10      1
Name: count, Length: 94, dtype: int64

In [21]:
# 2c. Check for missing values
print("Missing values per column:")
print(df.isnull().sum())
print(f"\nTotal missing: {df.isnull().sum().sum()}")

Missing values per column:
team_name                           0
year                                0
wins                                0
losses                              0
winning_percentage                  0
games_behind                      131
wild_card_games_behind           2028
record_in_the_last_10_games         0
current_streak                      0
runs_scored                         0
runs_allowed                        0
run_differential                    0
expected_win_loss_record            0
record_at_home                      0
record_when_away                    0
record_against_top_50_percent       0
dtype: int64

Total missing: 2159


### Your Exploration Findings

What did you learn from exploring the data? (Double-click this cell to edit)

**Example:** "My dataset has 15,234 rows and 8 columns. The 'borough' column has 5 unique values (Manhattan, Brooklyn, Queens, Bronx, Staten Island). The 'score' column ranges from 0 to 142, with a mean of 23.4. There are 847 missing values in the 'grade' column (about 5.5%)."

**Your answer:**

"My dataset has 2690 rows, 16 columns. The 'wins' column ranges from 10 to 116, with a mean of 77.There are 2028 missing values in the 'wild_card_games_behind' column. The Chicago Cubs' 1906 season had the most wins in a season of 116. The Chicago White Sox's 2024 had the most losses in a season of 121."

---
## Part 3: Handle Missing Values (5 points)
---

In [22]:
# 3a. Identify columns with missing values
missing_pct = (df.isnull().sum() / len(df)) * 100
print("Missing value percentages:")
print(missing_pct[missing_pct > 0])

Missing value percentages:
games_behind               4.869888
wild_card_games_behind    75.390335
dtype: float64


### Your Missing Value Decisions

For each column with missing values, explain what you did and why. (Double-click to edit)

**Options:**
- Drop rows: `df.dropna(subset=['column'])`
- Fill with a value: `df['column'].fillna('Unknown')`
- Fill with mean/median: `df['column'].fillna(df['column'].median())`
- Leave as-is (if not needed for your analysis)

**Example:**
- 'grade' column (5.5% missing): Dropped these rows because I need grades for my analysis and 5.5% is a small loss
- 'phone' column (12% missing): Left as-is because I don't need phone numbers for my research question

**Your decisions:**

wild_card_games_behind: I am filling with 0.0 rather than dropping. Since the Wild Card didn't exist before 1994, dropping them would delete 90+ years of baseball history.

games_behind: I am filling with 0.0. Missing values here usually represent the team in 1st place (who is 0 games behind).

In [23]:
# 3b. Implement your decisions
# 3b. Handle missing values without deleting history
df['games_behind'] = df['games_behind'].fillna(0.0)

df['wild_card_games_behind'] = df['wild_card_games_behind'].fillna(0.0)

# Verify that missing values are gone
print("Missing values after fix:")
print(df[['games_behind', 'wild_card_games_behind']].isnull().sum())

Missing values after fix:
games_behind              0
wild_card_games_behind    0
dtype: int64


---
## Part 4: Fix Data Types (5 points)
---

In [24]:
# 4a. Check current data types
df.dtypes

team_name                            str
year                               int64
wins                               int64
losses                             int64
winning_percentage               float64
games_behind                     float64
wild_card_games_behind           float64
record_in_the_last_10_games          str
current_streak                       str
runs_scored                        int64
runs_allowed                       int64
run_differential                   int64
expected_win_loss_record             str
record_at_home                       str
record_when_away                     str
record_against_top_50_percent        str
dtype: object

In [25]:
# 4b. Convert columns to appropriate types
# Common conversions:
#   Dates:      df['date'] = pd.to_datetime(df['date'])
#   Numbers:    df['col'] = df['col'].astype(int)
#   Categories: df['col'] = df['col'].astype('category')

df['year'] = pd.to_datetime(df['year'], format='%Y')
df['winning_percentage'] = df['winning_percentage'].astype(float)

# Check the results
print(df.dtypes)

team_name                                   str
year                             datetime64[us]
wins                                      int64
losses                                    int64
winning_percentage                      float64
games_behind                            float64
wild_card_games_behind                  float64
record_in_the_last_10_games                 str
current_streak                              str
runs_scored                               int64
runs_allowed                              int64
run_differential                          int64
expected_win_loss_record                    str
record_at_home                              str
record_when_away                            str
record_against_top_50_percent               str
dtype: object


### Your Type Conversions

**Example:**
- 'inspection_date': string -> datetime (for time-based analysis)
- 'zipcode': keeping as string (it's not really a number - 07024)

**Your conversions:**

(Write here)

In [26]:
# 4c. Check for problematic data

# 1. Check for duplicate rows (e.g., the same team in the same year appearing twice)
duplicate_count = df.duplicated(subset=['team_name', 'year']).sum()
print(f"Duplicate rows found: {duplicate_count}")

# 2. Look for invalid Winning Percentages (Should be between 0 and 1)
invalid_wins = df[(df['winning_percentage'] > 1.0) | (df['winning_percentage'] < 0)]
print(f"Invalid winning percentages: {len(invalid_wins)}")

# 3. Look for 'impossible' years (e.g., years before 1876 or after 2026)
invalid_years = df[(df['year'].dt.year < 1876) | (df['year'].dt.year > 2026)]
print(f"Invalid years found: {len(invalid_years)}")

Duplicate rows found: 0
Invalid winning percentages: 0
Invalid years found: 0


### Data Issues Found and Fixed

**Example:**
- Found 23 duplicate rows - removed them
- Found 3 restaurants with score = -1 (invalid) - removed those rows

**Your issues:**

- "Found 0 duplicate rows—the dataset is unique per team/year."

- "Confirmed all winning_percentage values are between 0.0 and 1.0."

- "Verified all season years fall within the professional era (1876–Present)."

---
## Part 5: Create Clean Dataset (5 points)
---

In [27]:
# 5a. Select only the columns you need for analysis
# key_columns = ['column1', 'column2', 'column3']
# df_clean = df[key_columns].copy()

# 10 target teams
historic_10 = [
    'Chicago Cubs', 'Los Angeles Dodgers', 'Atlanta Braves', 
    'San Francisco Giants', 'New York Yankees', 'Boston Red Sox', 
    'St. Louis Cardinals', 'Philadelphia Phillies', 'Pittsburgh Pirates', 
    'New York Mets'
]

# 2. Map historical names to modern franchises
franchise_map = {
    # Dodgers (Brooklyn -> LA)
    'Brooklyn Superbas': 'Los Angeles Dodgers', 
    'Brooklyn Robins': 'Los Angeles Dodgers', 
    'Brooklyn Dodgers': 'Los Angeles Dodgers',
    
    # Braves (Boston -> Milwaukee -> Atlanta)
    'Boston Beaneaters': 'Atlanta Braves', 'Boston Doves': 'Atlanta Braves',
    'Boston Rustlers': 'Atlanta Braves', 'Boston Braves': 'Atlanta Braves',
    'Boston Bees': 'Atlanta Braves', 'Milwaukee Braves': 'Atlanta Braves',
    
    # Giants, Yankees, and Cubs
    'New York Giants': 'San Francisco Giants',
    'New York Highlanders': 'New York Yankees',
    'Chicago Orphans': 'Chicago Cubs',
    
    # Red Sox
    'Boston Americans': 'Boston Red Sox'
}

# 3. Apply the mapping and fix spaces
df['team_name'] = df['team_name'].str.strip()
df['franchise'] = df['team_name'].replace(franchise_map)

# 4. Filter for only the Key Columns needed for "Better/Worse" trends
key_cols = ['franchise', 'team_name', 'year', 'wins', 'losses', 'winning_percentage', 'run_differential', 'runs_scored', 'runs_allowed']
df_clean = df[key_cols].copy()
df['year'] = pd.to_datetime(df['year'])
df_short = df_clean[(df_clean['franchise'].isin(historic_10)) & (df['year'].dt.year <= 2024)].copy()



In [28]:
# 5b. Verify the cleaned data
# print("CLEANED DATA SUMMARY")
# print("=" * 50)
# print(f"Shape: {df_clean.shape}")
# print(f"\nMissing values:\n{df_clean.isnull().sum()}")
# print(f"\nData types:\n{df_clean.dtypes}")
# df_clean.head()

print("CLEANED DATA SUMMARY")
print("=" * 50)
print(f"Shape: {df_short.shape}")
print(f"\nMissing values:\n{df_short.isnull().sum()}")
print(f"\nData types:\n{df_short.dtypes}")
df_short.head()

CLEANED DATA SUMMARY
Shape: (1168, 9)

Missing values:
franchise             0
team_name             0
year                  0
wins                  0
losses                0
winning_percentage    0
run_differential      0
runs_scored           0
runs_allowed          0
dtype: int64

Data types:
franchise                        str
team_name                        str
year                  datetime64[us]
wins                           int64
losses                         int64
winning_percentage           float64
run_differential               int64
runs_scored                    int64
runs_allowed                   int64
dtype: object


,franchise,team_name,year,wins,losses,winning_percentage,run_differential,runs_scored,runs_allowed
0,Pittsburgh Pirates,Pittsburgh Pirates,1901-01-01,90,49,0.647,242,776,534
2,Philadelphia Phillies,Philadelphia Phillies,1901-01-01,83,57,0.593,125,668,543
3,Boston Red Sox,Boston Americans,1901-01-01,79,57,0.581,151,759,608
4,Los Angeles Dodgers,Brooklyn Superbas,1901-01-01,79,57,0.581,144,744,600
7,St. Louis Cardinals,St. Louis Cardinals,1901-01-01,76,64,0.543,103,792,689


In [29]:
# 5c. Save the cleaned dataset
df_short.to_csv('outputs/cleaned_data.csv', index=False)
print("Cleaned dataset saved to 'cleaned_mlb_stats.csv'")

Cleaned dataset saved to 'cleaned_mlb_stats.csv'


---
## Cleaning Summary (required)
---

Fill in your cleaning process: (Double-click to edit)

1. **Original dataset:** 2690 rows, 16 columns
2. **Missing values handled:**
   - Column: wild_card_games_behind -> Action: filled with 0.0
   - Column: games_behind -> Action: filled with 0.0
3. **Data types fixed:**
   - Column: year -> Converted to: datetime
   - Column: winning_percentage -> Converted to: float
4. **Problems fixed:**
   - Issue: Francise moves/Rebrands -> Solution: Created franchise column to map historical names to modern teams.
   - Issue: Teams -> Solution: restricted the data to only be from 10 teams
   - Issue: Incorrect data in dataset -> Solution: omitted the 2025 season
5. **Final clean dataset:** 1169 rows, 9 columns
6. **Columns kept for analysis:**
   - franchise
   - team_name
   - year
   - wins 
   - losses
   - winning_percentage 
   - run_differential
   - runs_scored
   - runs_allowed

---
## Submission Checklist
---

Before submitting, verify:

- [ X ] My code runs without errors (Kernel > Restart & Run All)
- [ X ] I explored the data and documented findings
- [ X ] I handled all missing values (documented decisions)
- [ X ] I fixed data types where needed
- [ X ] I removed duplicates and invalid data
- [ X ] I saved cleaned data to outputs/cleaned_data.csv
- [ X ] I filled in the cleaning summary

**Submit:** This notebook + your original data file + your cleaned data file